File: ruby/liquidplanner.rb
#!/usr/bin/env ruby

# You'll need the httparty (and highline, for the demo) gem
#
require 'httparty'

class LiquidPlanner
  include HTTParty

  base_uri 'https://app.liquidplanner.com/api'
  format :json

  attr_accessor :workspace_id

  def initialize(email, password)
    @opts = { :basic_auth => { :username => email, :password => password },
              :headers    => { 'content-type' => 'application/json' },
            }
  end

  def get(url, options={})
    self.class.get(url, options.merge(@opts))
  end

  def post(url, options={})
    options[:body] = options[:body].to_json if options[:body]
    self.class.post(url, options.merge(@opts))
  end

  def put(url, options={})
    options[:body] = options[:body].to_json if options[:body]
    self.class.put(url, options.merge(@opts))
  end

  def account
    get('/account')
  end

  def workspaces
    get('/workspaces')
  end

  def projects
    get("/workspaces/#{workspace_id}/projects")
  end

  def tasks
    get("/workspaces/#{workspace_id}/tasks")
  end

  def create_task(data)
    options = { :body => { :task => data } }
    post("/workspaces/#{workspace_id}/tasks", :body => { :task => data })
  end

  def update_task(data)
    options = { :body => { :task => data } }
    put("/workspaces/#{workspace_id}/tasks/#{data['id']}", :body => { :task => data })
  end

  def self.demo
    email    = ask("LiquidPlanner email: ")
    password = ask("LiquidPlanner password for #{email}: ") {|q| q.echo = false}

    lp = LiquidPlanner.new(email, password)

    account = lp.account
    puts "You are #{account['user_name']} (#{account['id']})"

    workspaces = lp.workspaces
    puts "You have #{workspaces.length} workspace#{workspaces.length == 1 ? '' : 's'}"
    workspaces.each {|ws| puts " #{ws['name']}"}

    ws = workspaces.first
    lp.workspace_id = ws['id']

    projects = lp.projects
    puts "These are the #{projects.length} projects in your '#{ws['name']}' workspace"
    projects.each_with_index do |p, i|
      puts " #{i+1}. #{p['name']}"
    end

    add_task = ask "Should I add a task to your first project? (y for yes) "
    if 'Y' == add_task.strip[0].upcase
      task = { 'name' => 'learn the API', 'parent_id' => projects.first['id'] }
      result = lp.create_task(task)
      update = { 'name' => 'learn more about the API', 'id' => result['id'] } 
      puts lp.update_task(update).inspect
    end

  end

end

if $0 == __FILE__
  require 'highline/import'
  LiquidPlanner.demo
end

File: php/liquidplanner.php
<?php
// This is an example of how to use the LiquidPlanner API in PHP.
class LiquidPlanner {
  private $_base_uri = "https://app.liquidplanner.com/api";
  private $_ch;
  public  $workspace_id;

  function __construct($email, $password) {
    $this->_ch = curl_init();
    curl_setopt($this->_ch, CURLOPT_HEADER, false);
    curl_setopt($this->_ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($this->_ch, CURLOPT_USERPWD, "$email:$password");
    curl_setopt($this->_ch, CURLOPT_HTTPHEADER, array('content-type: application/json'));
    curl_setopt($this->_ch, CURLOPT_ENCODING, 'gzip');
  }

  public function get($url) {
    curl_setopt($this->_ch, CURLOPT_HTTPGET, true);
    curl_setopt($this->_ch, CURLOPT_URL, $this->_base_uri.$url);
    return json_decode(curl_exec($this->_ch));
  }

  public function post($url, $body=null) {
    curl_setopt($this->_ch, CURLOPT_POST, true);
    curl_setopt($this->_ch, CURLOPT_URL, $this->_base_uri.$url);
    curl_setopt($this->_ch, CURLOPT_POSTFIELDS, json_encode($body));
    return json_decode(curl_exec($this->_ch));
  }

  public function put($url, $body=null) {
    curl_setopt($this->_ch, CURLOPT_CUSTOMREQUEST, 'PUT');
    curl_setopt($this->_ch, CURLOPT_URL, $this->_base_uri.$url);
    curl_setopt($this->_ch, CURLOPT_POSTFIELDS, json_encode($body));
    return json_decode(curl_exec($this->_ch));
  }

  public function account() {
    return $this->get('/account');
  }

  public function workspaces() {
    return $this->get('/workspaces');
  }

  public function projects() {
    return $this->get("/workspaces/{$this->workspace_id}/projects");
  }

  public function tasks() {
    return $this->get("/workspaces/{$this->workspace_id}/tasks");
  }

  public function create_task($data) {
    return $this->post("/workspaces/{$this->workspace_id}/tasks", array("task"=>$data));
  }

  public function update_task($data) {
    return $this->put("/workspaces/{$this->workspace_id}/tasks/{$data['id']}", array("task"=>$data));
  }

  public static function demo() {
    echo "LiquidPlanner email: ";
    $email = trim(fgets(STDIN));
    echo "LiquidPlanner password for $email: ";
    system('stty -echo');
    $password = trim(fgets(STDIN));
    system('stty echo');
    echo PHP_EOL;
    
    $lp = new LiquidPlanner($email, $password);
    
    $account = $lp->account();
    echo "You are $account->user_name ($account->id)".PHP_EOL;
    
    $workspaces = $lp->workspaces();
    $count = count($workspaces);
    $plural = $count == 1 ? '' : 's';
    echo "You have $count workspace$plural".PHP_EOL;
    foreach($workspaces as $ws) {
      echo " $ws->name\n";
    }

    $ws = $workspaces[0];
    $lp->workspace_id = $ws->id;

    $projects = $lp->projects();
    $count = count($projects);
    echo "These are the $count projects in your '$ws->name' workspace".PHP_EOL;
    foreach($projects as $i => $p) {
      echo ' '.($i+1).'. '.$p->name.PHP_EOL;
    }

    echo "Should I add a task to your first project? (y for yes) ";
    $add_task = fgets(STDIN);
    if('Y' == strtoupper(substr(trim($add_task), 0, 1))) {
      $task = array( 'name' => 'learn the API', 'parent_id' => $projects[0]->id );
      $result = $lp->create_task($task);
      $update = array( 'name' => 'learn more about the API', 'id' => $result->id );
      print_r($lp->update_task($update));
    }

  }

}

if ($argv[0] == 'liquidplanner.php') {
  LiquidPlanner::demo();
}

?>
